java-checks/src/main/java/org/sonar/java/checks/SubClassReferenceInitalizationCheck.java

```diff
@@ -0,0 +1,100 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+@Rule(key = "S2390")
+public class SubClassReferenceInitalizationCheck extends IssuableSubscriptionVisitor {
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    ClassTree classTree = (ClassTree) tree;
+    Type classType = classTree.symbol().type();
+    List<Tree> members = classTree.members();
+
+    checkVariables(members, classType);
+    checkStaticInitilizers(members, classType);
+  }
+
+  private void checkVariables(List<Tree> members, Type classType) {
+    members.stream()
+      .filter(member -> member.is(Tree.Kind.VARIABLE))
+      .map(VariableTree.class::cast)
+      .filter(SubClassReferenceInitalizationCheck::isStaticVariable)
+      .map(VariableTree::initializer)
+      .filter(Objects::nonNull)
+      .forEach(initializer -> initializer.accept(new StaticAccessVisitor(classType)));
+  }
+
+  private static boolean isStaticVariable(VariableTree tree) {
+    return tree.symbol().isStatic();
+  }
+
+  private void checkStaticInitilizers(List<Tree> members, Type classType) {
+    members.stream()
+      .filter(member -> member.is(Tree.Kind.STATIC_INITIALIZER))
+      .forEach(tree -> tree.accept(new StaticAccessVisitor(classType)));
+  }
+
+
+  private class StaticAccessVisitor extends BaseTreeVisitor {
+    private final Type classType;
+
+    public StaticAccessVisitor(Type classType) {
+      this.classType = classType;
+    }
+
+    @Override
+    public void visitAssignmentExpression(AssignmentExpressionTree tree) {
+      // skip the variable
+      scan(tree.expression());
+    }
+
+    @Override
+    public void visitIdentifier(IdentifierTree tree) {
+      Type type = tree.symbolType();
+      if (!classType.equals(type) && type.isSubtypeOf(classType)) {
```

Comment:
If I'm correct this will raise an issue for anonymous classes ?
```
class Foo {
static Foo foo = new Foo() { // issue ?
 .. . .
   FooSubclass fun() { // issue
     ... 
   }
}
```